File: backend/physics/integrators.py
from __future__ import annotations
import torch
from .particle import ParticleSystem


def verlet_integrate(ps: ParticleSystem, dt: float) -> None:
    """Verlet integration matching traerphysics.js Particle.integrate():
        vel = pos - prev
        next = pos + vel + acc * dt²
        prev = pos
    """
    mask = ~ps.pinned.unsqueeze(-1)  # (N, 1)
    dt_sq = dt * dt

    current = ps.positions.clone()
    velocity = current - ps.prev_positions
    ps.positions = torch.where(
        mask,
        current + velocity + ps.accelerations * dt_sq,
        current,
    )
    ps.prev_positions = current
    ps.accelerations.zero_()


def euler_integrate(ps: ParticleSystem, dt: float) -> None:
    """Explicit Euler integration for modes needing explicit velocity (mesh3d collisions).
        v += a * dt
        x += v * dt
    """
    mask = ~ps.pinned.unsqueeze(-1)  # (N, 1)

    ps.velocities += ps.accelerations * dt
    ps.velocities = torch.where(mask, ps.velocities, torch.zeros_like(ps.velocities))
    ps.positions += ps.velocities * dt * mask.float()
    ps.accelerations.zero_()


def clamp_velocities(ps: ParticleSystem, max_speed: float) -> None:
    """Clamp particle velocities to a maximum speed."""
    if max_speed <= 0:
        return
    speed = ps.velocities.norm(dim=-1, keepdim=True)  # (N, 1)
    too_fast = speed > max_speed
    scale = torch.where(too_fast, max_speed / speed.clamp(min=1e-6), torch.ones_like(speed))
    ps.velocities *= scale

File: frontend/src/physics/integrators.ts
import type { ParticleSystem } from "./ParticleSystem";

/**
 * Verlet integration matching traerphysics.js Particle.integrate():
 *   vel = pos - prev
 *   next = pos + vel + acc * dt²
 *   prev = pos
 */
export function verletIntegrate(ps: ParticleSystem, dt: number): void {
  const dim = ps.dim;
  const pos = ps.positions;
  const prev = ps.prevPositions;
  const acc = ps.accelerations;
  const dtSq = dt * dt;

  for (let i = 0; i < ps.count; i++) {
    if (ps.pinned[i]) {
      for (let d = 0; d < dim; d++) {
        acc[i * dim + d] = 0;
      }
      continue;
    }
    for (let d = 0; d < dim; d++) {
      const idx = i * dim + d;
      const current = pos[idx];
      const velocity = current - prev[idx];
      pos[idx] = current + velocity + acc[idx] * dtSq;
      prev[idx] = current;
      acc[idx] = 0;
    }
  }
}

/**
 * Explicit Euler integration (for mesh3d where explicit velocity is needed for collisions).
 *   v += a * dt
 *   x += v * dt
 */
export function eulerIntegrate(ps: ParticleSystem, dt: number): void {
  const dim = ps.dim;
  const pos = ps.positions;
  const vel = ps.velocities;
  const acc = ps.accelerations;

  for (let i = 0; i < ps.count; i++) {
    if (ps.pinned[i]) {
      for (let d = 0; d < dim; d++) {
        acc[i * dim + d] = 0;
        vel[i * dim + d] = 0;
      }
      continue;
    }
    for (let d = 0; d < dim; d++) {
      const idx = i * dim + d;
      vel[idx] += acc[idx] * dt;
      pos[idx] += vel[idx] * dt;
      acc[idx] = 0;
    }
  }
}

export function clampVelocities(ps: ParticleSystem, maxSpeed: number): void {
  if (maxSpeed <= 0) return;
  const dim = ps.dim;
  const vel = ps.velocities;

  for (let i = 0; i < ps.count; i++) {
    let speedSq = 0;
    for (let d = 0; d < dim; d++) {
      speedSq += vel[i * dim + d] * vel[i * dim + d];
    }
    if (speedSq > maxSpeed * maxSpeed) {
      const scale = maxSpeed / Math.sqrt(speedSq);
      for (let d = 0; d < dim; d++) {
        vel[i * dim + d] *= scale;
      }
    }
  }
}
